### PR TITLE
Support for managing EC2 Volumes

### PR DESCRIPTION
**Pull request by daveseff**
I am adding support for managing EC2 Block device volumes. With this you can create and attach volumes to already running instances as well as detach and destroy.

Volumes can be created from scratch or built from a snapshot.

### CODE REVIEW

File: spec/unit/provider/ec2_volume/v2_spec.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+require 'spec_helper'
+
+provider_class = Puppet::Type.type(:ec2_volume).provider(:v2)
+
+ENV['AWS_ACCESS_KEY_ID'] = 'redacted'
+ENV['AWS_SECRET_ACCESS_KEY'] = 'redacted'
+ENV['AWS_REGION'] = 'sa-east-1'
```
Comment (zleswomp) [inline spec/unit/provider/ec2_volume/v2_spec.rb]: These variables won't need to exist here if https://github.com/puppetlabs/puppetlabs-aws/pull/399 is merged.

**Comment by zleswomp** [comment]
This code looks good to me.  Thank you for putting in the effort!

File: README.md [hunk 2]
```diff
@@ -528,6 +529,35 @@ back- end instances.  Accepts a hash with the following keys:
...
+*Required* The volume type. This can be gp2 for General Purpose SSD, io1 for Provisioned IOPS SSD, st1 for Throughput Optimized HDD, sc1 for Cold HDD, or standard for Magnetic volumes.
+
+##### `encrypted`
+*Optional* Specifies whether the volume should be encrypted. Encrypted Amazon EBS volumes may only be attached to instances that support Amazon EBS encryption. Volumes that are created from encrypted snapshots are automatically encrypted. There is no way to create an encrypted volume from an unencrypted snapshot or vice versa.
+
+##### `kms_key_id`
+*Optional* The full ARN of the AWS Key Management Service (AWS KMS) customer master key (CMK) to use when creating the encrypted volume. This parameter is only required if you want to use a non-default CMK; if this parameter is not specified, the default CMK for EBS is used.
+
+##### `snapshot_id`
+*Optional* The snapshot from which to create the volume.
```
Comment (zabacad) [inline README.md]: This should be *Conditional* to match *size*.

File: lib/puppet/provider/ec2_volume/v2.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  def destroy
+    Puppet.info("Deleting Volume #{name} in region #{target_region}")
+    # Detach if in use first
+    config = {
+      volume_id: volume_id,
+      force: true
+    }
+    ec2.detach_volume(config) unless @property_hash[:attach] == nil
+    ec2.wait_until(:volume_available, volume_ids: [volume_id])
```
Comment (zabacad) [inline lib/puppet/provider/ec2_volume/v2.rb]: Waiting could cause a very long run. I prefer the pattern of attempting the deletion immediately, and if it fails, notify the user to run Puppet again. Check the `cloudfront_distribution` `destroy` method for an example.

**Comment by zabacad** [comment]
None of my comments should stand against a merge since this is a new resource type with no established behaviour.

File: lib/puppet/provider/ec2_volume/v2.rb [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    @property_hash[:id] = response.volume_id
+    @property_hash[:ensure] = :present
+  end
+
+  def destroy
+    Puppet.info("Deleting Volume #{name} in region #{target_region}")
+    # Detach if in use first
+    config = {
+      volume_id: volume_id,
+      force: true
```
Comment (daveseff) [inline lib/puppet/provider/ec2_volume/v2.rb]: This is in the destroy method of the module. The intention is that this is a volume to be deleted with prejudice. It's actually sparks an idea to have a parameter 'ensure => detached' for those volumes that should be detached with data preserved. If this is a desired feature I can look at that in the future.

Comment (zabacad) [inline lib/puppet/provider/ec2_volume/v2.rb]: Okay, since the volume is immediately deleted, there's no reason to care about its data.

**Comment by zleswomp** [comment]
I'm still +1 here.  Merging.

**Pull request merged**